### PR TITLE
SNOW-1830203: Add a new parameter for CTE optimization 

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1830203

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   We need a new parameter for control snowpark version

### CODE REVIEW

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -558,11 +560,17 @@ def __init__(
                 _PYTHON_SNOWPARK_USE_SQL_SIMPLIFIER_STRING, True
             )
         )
-        self._cte_optimization_enabled: bool = (
+        cte_optimization_enabled_version = (
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/session.py:563]: Why only the parameters for cte optimization and temp table auto cleanup depend on snowpack's version but not the other params (e.g., describe queries, large query breakdown, etc)?

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM but had one question.

File: src/snowflake/snowpark/session.py (hunk 1 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py:563]: yes other parameters will also be changed in this way later!

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -558,11 +560,17 @@ def __init__(
...
             self._conn._get_client_side_session_parameter(
-                _PYTHON_SNOWPARK_USE_CTE_OPTIMIZATION_STRING, False
+                _PYTHON_SNOWPARK_USE_CTE_OPTIMIZATION_VERSION, ""
             )
         )
+        self._cte_optimization_enabled: bool = (
+            isinstance(cte_optimization_enabled_version, str)
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/session.py:569]: @sfc-gh-jdu it seems we will have couple of parameters to perform similar logic, how about extract a common utility function,
 is_feature_enabled_for_version(feature_version_parameter), the function will get the version configured for the feature, and check if the current version >= the configured version

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:569]: +1

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py]: good point! Added such a function

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/_internal/utils.py [hunk 3]
```diff
@@ -1280,3 +1281,19 @@ def import_or_missing_modin_pandas() -> Tuple[ModuleLikeObject, bool]:
         return modin, True
     except ImportError:
         return MissingModin(), False
+
+
+def is_feature_enabled_for_version(
+    session: "snowflake.snowpark.Session", parameter_name: str
+) -> bool:
+    """
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/utils.py:1289]: that doesn't look like a proper utility funciton to put in _internal/utils.py, maybe just addd this as an internal function for session.py

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/utils.py]: good point, fixed

**Review by sfc-gh-yzou — APPROVED** [review]

**Pull request merged**